qutebrowser/completion/completionwidget.py

```diff
@@ -116,6 +117,7 @@ def __init__(self, *,
                  parent: QWidget = None) -> None:
         super().__init__(parent)
         self.pattern = None  # type: typing.Optional[str]
+        self._model_pattern = None  # type: typing.Optional[str]
```

Comment:
Ignore ^^^ ... I forgot how `completion.show=auto` worked :laughing: 

I'm starting to think our model of signals/events here is wrong. We're operating on the idea that typing sends a signal to update the completion (probably because I wrote this and I only use `show=always` :facepalm:).

When `show=auto`, we should let `TAB` trigger setting the completion, rather than triggering on every key press and trying to decide if we should handle it.

Could we maybe add something to https://github.com/qutebrowser/qutebrowser/blob/5baeb28ab652831f726001c484a2a726543691f5/qutebrowser/completion/completionwidget.py#L185-L192 that calls `set_pattern` on the model if `show=auto`?

Then here we only call `set_pattern` if `show=always`? https://github.com/qutebrowser/qutebrowser/blob/5baeb28ab652831f726001c484a2a726543691f5/qutebrowser/completion/completionwidget.py#L382-L387

Maybe this is more confusing, in which case ignore me :)

Reply:
I like that idea @rcorre. Definitely sounds better that having a `set_pattern` method that doesn't set a pattern because of some external setting.

Reply:
Wait, no, I misunderstood what you said. I think what you propose is more or less what I did, only that I did it in `completion_item_focus` instead of `_next_idx`.

Reply:
ah, so you are. I was thinking that `completion_item_focus()` could update the pattern only if show == auto and we're not visible (since you should only have to update it when newly showing the menu). That wouldn't work for `never` though, so I guess we need to cache a second pattern.

Reply:
> Or maybe that logic should be in `CompletionModel` entirely, so that calling `set_pattern` on it just is a no-op if the pattern is already the same?
I followed this approach now and it looks much better. Thanks.